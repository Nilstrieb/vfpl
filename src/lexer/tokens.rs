use crate::error::Span;
use std::fmt::{Display, Formatter};

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub span: Span,
    pub kind: TokenType,
}

#[derive(Debug, Clone, PartialEq)]
pub enum TokenType {
    Please,

    Ident(String),
    // Keywords
    As,
    Add,
    Subtract,
    Multiply,
    Divide,
    Modulo,
    Initialize,
    Variable,
    With,
    The,
    Value,
    Of,
    Set,
    To,
    From,
    By,
    Take,
    End,
    Check,
    Whether,
    Then,
    Do,
    Otherwise,
    Break,
    Out,
    This,
    Create,
    Function,
    Parameter,
    Parameters,
    That,
    Returns,
    Call,
    No,
    Argument,
    Arguments,
    Go,
    Sleep,
    Absent,
    Null,
    NoValue,
    Undefined,
    True,
    False,
    Does,
    Has,
    Is,

    // literals
    String(String),
    Int(i64),
    Float(f64),

    // symbols
    /// .
    Dot,
    /// (
    ParenOpen,
    /// )
    ParenClose,
    /// ,
    Comma,
}

impl Display for TokenType {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            TokenType::Please => f.write_str("keyword `please`"),
            TokenType::Ident(name) => f.write_str(&format!("`{}`", name)),
            TokenType::As => f.write_str("keyword `as`"),
            TokenType::Add => f.write_str("keyword `add`"),
            TokenType::Subtract => f.write_str("keyword `subtract`"),
            TokenType::Multiply => f.write_str("keyword `multiply`"),
            TokenType::Divide => f.write_str("keyword `divide`"),
            TokenType::Modulo => f.write_str("keyword `modulo`"),
            TokenType::Initialize => f.write_str("keyword `initialize`"),
            TokenType::Variable => f.write_str("keyword `variable`"),
            TokenType::With => f.write_str("keyword `with`"),
            TokenType::The => f.write_str("keyword `the`"),
            TokenType::Value => f.write_str("keyword `value`"),
            TokenType::Of => f.write_str("keyword `of`"),
            TokenType::Set => f.write_str("keyword `set`"),
            TokenType::To => f.write_str("keyword `to`"),
            TokenType::From => f.write_str("keyword `from`"),
            TokenType::By => f.write_str("keyword `by`"),
            TokenType::Take => f.write_str("keyword `take`"),
            TokenType::End => f.write_str("keyword `end`"),
            TokenType::Check => f.write_str("keyword `check`"),
            TokenType::Whether => f.write_str("keyword `whether`"),
            TokenType::Then => f.write_str("keyword `then`"),
            TokenType::Do => f.write_str("keyword `do`"),
            TokenType::Otherwise => f.write_str("keyword `otherwise`"),
            TokenType::Break => f.write_str("keyword `break`"),
            TokenType::Out => f.write_str("keyword `out`"),
            TokenType::This => f.write_str("keyword `this`"),
            TokenType::Create => f.write_str("keyword `create`"),
            TokenType::Function => f.write_str("keyword `function`"),
            TokenType::Parameter => f.write_str("keyword `parameter`"),
            TokenType::Parameters => f.write_str("keyword `parameters`"),
            TokenType::That => f.write_str("keyword `that`"),
            TokenType::Returns => f.write_str("keyword `returns`"),
            TokenType::Go => f.write_str("keyword `go`"),
            TokenType::Sleep => f.write_str("keyword `sleep`"),
            TokenType::Absent => f.write_str("keyword `absent`"),
            TokenType::Null => f.write_str("keyword `null`"),
            TokenType::NoValue => f.write_str("keyword `no value` )"),
            TokenType::Undefined => f.write_str("keyword `undefined`"),
            TokenType::True => f.write_str("keyword `true`"),
            TokenType::False => f.write_str("keyword `false`"),
            TokenType::String(value) => f.write_str(&format!("`\"{}\"`", value)),
            TokenType::Int(value) => f.write_str(&format!("`{}`", value)),
            TokenType::Float(value) => f.write_str(&format!("`{}`", value)),
            TokenType::Dot => f.write_str("`.`"),
            TokenType::ParenOpen => f.write_str("`(`"),
            TokenType::ParenClose => f.write_str("`)`"),
            TokenType::Comma => f.write_str("`,`"),
        }
    }
}
