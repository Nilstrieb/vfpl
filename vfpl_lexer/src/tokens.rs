use std::fmt::{Display, Formatter};
use vfpl_error::Span;

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub span: Span,
    pub kind: TokenKind,
}

#[derive(Debug, Clone, PartialEq)]
pub enum TokenKind {
    CondKeyword(CondKeyword),
    Ident(String),
    // Keywords
    Please,
    As,
    Initialize,
    Variable,
    End,
    Check,
    Whether,
    Then,
    Do,
    Otherwise,
    Break,
    This,
    Create,
    Function,
    Call,
    And,
    Absent,
    Null,
    NoValue,
    Undefined,
    True,
    False,
    Not,
    Or,
    Repeat,
    Return,
    While,

    // literals
    String(String),
    Int(i64),
    Float(f64),

    // symbols
    /// .
    Dot,
    /// (
    ParenOpen,
    /// )
    ParenClose,
    /// ,
    Comma,

    /// The last token
    Eof,
}

/// A conditional keyword that can be used as an identifier
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CondKeyword {
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    With,
    The,
    Value,
    Of,
    Set,
    To,
    From,
    By,
    Take,
    Out,
    Parameter,
    Parameters,
    That,
    Returns,
    No,
    Argument,
    Arguments,
    Go,
    Sleep,
    Does,
    Has,
    Is,
    Have,
    Greater,
    Less,
    Than,
    Equal,
}

impl CondKeyword {
    fn as_str(&self) -> &'static str {
        match self {
            CondKeyword::Add => "add",
            CondKeyword::Sub => "sub",
            CondKeyword::Mul => "mul",
            CondKeyword::Div => "div",
            CondKeyword::Mod => "mod",
            CondKeyword::With => "with",
            CondKeyword::The => "the",
            CondKeyword::Value => "value",
            CondKeyword::Of => "of",
            CondKeyword::Set => "set",
            CondKeyword::To => "to",
            CondKeyword::From => "from",
            CondKeyword::By => "by",
            CondKeyword::Take => "take",
            CondKeyword::Out => "out",
            CondKeyword::Parameter => "parameter",
            CondKeyword::Parameters => "parameters",
            CondKeyword::That => "that",
            CondKeyword::Returns => "returns",
            CondKeyword::No => "no",
            CondKeyword::Argument => "argument",
            CondKeyword::Arguments => "arguments",
            CondKeyword::Go => "go",
            CondKeyword::Sleep => "sleep",
            CondKeyword::Does => "does",
            CondKeyword::Has => "has",
            CondKeyword::Is => "is",
            CondKeyword::Have => "have",
            CondKeyword::Greater => "greater",
            CondKeyword::Less => "less",
            CondKeyword::Than => "than",
            CondKeyword::Equal => "equal",
        }
    }
}

impl Display for TokenKind {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            TokenKind::Please => f.write_str("keyword `please`"),
            TokenKind::Ident(name) => f.write_str(&format!("`{}`", name)),
            TokenKind::As => f.write_str("keyword `as`"),
            TokenKind::Initialize => f.write_str("keyword `initialize`"),
            TokenKind::Variable => f.write_str("keyword `variable`"),
            TokenKind::End => f.write_str("keyword `end`"),
            TokenKind::Check => f.write_str("keyword `check`"),
            TokenKind::Whether => f.write_str("keyword `whether`"),
            TokenKind::Then => f.write_str("keyword `then`"),
            TokenKind::Do => f.write_str("keyword `do`"),
            TokenKind::Otherwise => f.write_str("keyword `otherwise`"),
            TokenKind::Break => f.write_str("keyword `break`"),
            TokenKind::This => f.write_str("keyword `this`"),
            TokenKind::Create => f.write_str("keyword `create`"),
            TokenKind::Function => f.write_str("keyword `function`"),
            TokenKind::Absent => f.write_str("keyword `absent`"),
            TokenKind::Null => f.write_str("keyword `null`"),
            TokenKind::NoValue => f.write_str("keyword `novalue` )"),
            TokenKind::Undefined => f.write_str("keyword `undefined`"),
            TokenKind::True => f.write_str("keyword `true`"),
            TokenKind::False => f.write_str("keyword `false`"),
            TokenKind::Not => f.write_str("keyword `not`"),
            TokenKind::Or => f.write_str("keyword `or`"),
            TokenKind::Call => f.write_str("keyword `call`"),
            TokenKind::Repeat => f.write_str("keyword `repeat`"),
            TokenKind::Return => f.write_str("keyword `return`"),
            TokenKind::String(value) => f.write_str(&format!("`\"{}\"`", value)),
            TokenKind::Int(value) => f.write_str(&format!("`{}`", value)),
            TokenKind::Float(value) => f.write_str(&format!("`{}`", value)),
            TokenKind::Dot => f.write_str("`.`"),
            TokenKind::ParenOpen => f.write_str("`(`"),
            TokenKind::ParenClose => f.write_str("`)`"),
            TokenKind::Comma => f.write_str("`,`"),
            TokenKind::And => f.write_str("keyword `and`"),
            TokenKind::While => f.write_str("keyword `while`"),
            TokenKind::Eof => f.write_str("end of file"),
            TokenKind::CondKeyword(cond) => f.write_str(&format!("keyword `{}`", cond.as_str())),
        }
    }
}

impl Token {
    /// Token with a Span of one character
    pub fn new_from_single(kind: TokenKind, idx: usize) -> Self {
        Token {
            kind,
            span: Span::single(idx),
        }
    }

    pub fn new(kind: TokenKind, start: usize, end: usize) -> Self {
        Token {
            kind,
            span: Span::start_end(start, end),
        }
    }

    pub fn new_from_len(kind: TokenKind, start: usize, end: usize) -> Self {
        Token {
            kind,
            span: Span::start_len(start, end),
        }
    }

    pub fn eof() -> Self {
        Token {
            span: Span::eof(),
            kind: TokenKind::Eof,
        }
    }
}
