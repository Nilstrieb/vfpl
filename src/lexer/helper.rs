use crate::lexer::tokens::TokenKind;

pub fn compute_keyword(identifier: &str) -> Option<TokenKind> {
    Some(match identifier {
        "please" => TokenKind::Please,
        "as" => TokenKind::As,
        "add" => TokenKind::Add,
        "sub" => TokenKind::Sub,
        "mul" => TokenKind::Mul,
        "div" => TokenKind::Div,
        "modulo" => TokenKind::Mod,
        "initialize" => TokenKind::Initialize,
        "variable" => TokenKind::Variable,
        "with" => TokenKind::With,
        "the" => TokenKind::The,
        "value" => TokenKind::Value,
        "of" => TokenKind::Of,
        "set" => TokenKind::Set,
        "to" => TokenKind::To,
        "from" => TokenKind::From,
        "by" => TokenKind::By,
        "take" => TokenKind::Take,
        "end" => TokenKind::End,
        "check" => TokenKind::Check,
        "whether" => TokenKind::Whether,
        "then" => TokenKind::Then,
        "do" => TokenKind::Do,
        "otherwise" => TokenKind::Otherwise,
        "break" => TokenKind::Break,
        "out" => TokenKind::Out,
        "this" => TokenKind::This,
        "create" => TokenKind::Create,
        "function" => TokenKind::Function,
        "parameter" => TokenKind::Parameter,
        "parameters" => TokenKind::Parameters,
        "that" => TokenKind::That,
        "returns" => TokenKind::Returns,
        "call" => TokenKind::Call,
        "no" => TokenKind::No,
        "argument" => TokenKind::Argument,
        "arguments" => TokenKind::Arguments,
        "and" => TokenKind::And,
        "go" => TokenKind::Go,
        "sleep" => TokenKind::Sleep,
        "absent" => TokenKind::Absent,
        "null" => TokenKind::Null,
        "noValue" => TokenKind::NoValue,
        "undefined" => TokenKind::Undefined,
        "true" => TokenKind::True,
        "false" => TokenKind::False,
        "does" => TokenKind::Does,
        "has" => TokenKind::Has,
        "is" => TokenKind::Is,
        "not" => TokenKind::Not,
        "have" => TokenKind::Have,
        "greater" => TokenKind::Greater,
        "less" => TokenKind::Less,
        "than" => TokenKind::Than,
        "equal" => TokenKind::Equal,
        "or" => TokenKind::Or,
        "repeat" => TokenKind::Repeat,
        "return" => TokenKind::Return,
        "while" => TokenKind::While,
        _ => return None,
    })
}
