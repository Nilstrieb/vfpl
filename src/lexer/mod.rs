use std::option::Option::Some;
use std::str::CharIndices;

use peekmore::{PeekMore, PeekMoreIterator};
use unicode_xid::UnicodeXID;

use tokens::Token;

use crate::error::{CompilerError, Span};
use crate::lexer::helper::compute_keyword;
use crate::lexer::tokens::TokenKind;

mod helper;
#[cfg(test)]
mod test;
pub mod tokens;

pub type LexerResult<T> = Result<T, LexerError>;

#[derive(Debug)]
pub struct Lexer<'a> {
    char_indices: PeekMoreIterator<CharIndices<'a>>,
}

impl Lexer<'_> {
    pub fn next(&mut self) -> Option<(usize, char)> {
        self.char_indices.next()
    }

    /// Consume elements n times
    pub fn consume_elements(&mut self, n: usize) {
        for _ in 0..n {
            self.next();
        }
    }

    pub fn peek(&mut self) -> Option<(usize, char)> {
        self.char_indices.peek().copied()
    }

    pub fn peek_nth(&mut self, n: usize) -> Option<(usize, char)> {
        self.char_indices.peek_nth(n).copied()
    }

    pub fn new(str: &str) -> Lexer {
        Lexer {
            char_indices: str.char_indices().peekmore(),
        }
    }

    pub fn compute_tokens(&mut self) -> LexerResult<Vec<Token>> {
        let mut tokens = Vec::new();
        while let Some((idx, char)) = self.next() {
            match char {
                '.' => tokens.push(Token::new_from_single(TokenKind::Dot, idx)),
                '(' => tokens.push(Token::new_from_single(TokenKind::ParenOpen, idx)),
                ')' => tokens.push(Token::new_from_single(TokenKind::ParenClose, idx)),
                ',' => tokens.push(Token::new_from_single(TokenKind::Comma, idx)),
                '\"' => tokens.push(self.compute_string(idx)?),
                '<' => {
                    if self.peek_nth(0).ok_or(LexerError::UnexpectedEOF)?.1 == '!'
                        && self.peek_nth(1).ok_or(LexerError::UnexpectedEOF)?.1 == '-'
                        && self.peek_nth(2).ok_or(LexerError::UnexpectedEOF)?.1 == '-'
                    {
                        while let Some((_, char)) = self.next() {
                            if char == '-' {
                                if self.peek_nth(0).ok_or(LexerError::UnexpectedEOF)?.1 == '-'
                                    && self.peek_nth(1).ok_or(LexerError::UnexpectedEOF)?.1 == '>'
                                {
                                    self.consume_elements(2);
                                    break;
                                }
                            } else {
                                continue;
                            }
                        }
                    } else {
                        return Err(LexerError::InvalidCharacter(Span::single(idx)));
                    }
                }
                '0'..='9' | '-' => tokens.push(self.compute_number(char, idx)?),
                other if other.is_whitespace() => continue,
                other if other.is_xid_start() => tokens.push(self.compute_identifier(char, idx)),
                _ => return Err(LexerError::InvalidCharacter(Span::single(idx))),
            }
        }

        tokens.push(Token::eof());
        Ok(tokens)
    }

    fn compute_identifier(&mut self, char: char, idx: usize) -> Token {
        let mut identifier = String::from(char);
        while let Some((_, char)) = self.peek() {
            if char.is_xid_continue() {
                self.next();
                identifier.push(char);
            } else {
                break;
            }
        }
        identifier = identifier.to_lowercase();
        let end = identifier.len();
        let kind = compute_keyword(&identifier).unwrap_or(TokenKind::Ident(identifier));
        Token::new_from_len(kind, idx, end)
    }

    fn compute_number(&mut self, char: char, idx: usize) -> LexerResult<Token> {
        let mut number = String::from(char);
        let mut end = loop {
            if let Some((i, char)) = self.peek() {
                match char {
                    '0'..='9' => {
                        self.next();
                        number.push(char);
                    }
                    other if other.is_xid_start() => {
                        return Err(LexerError::InvalidCharacter(Span::single(i)));
                    }
                    _ => {
                        // This works because the character before it will be 1 byte long (0..9 || -)
                        break i;
                    }
                }
            } else {
                break idx + number.len();
            }
        };

        if let Some((i, char)) = self.peek() {
            if char == '.' {
                if let Some((i, char)) = self.peek_nth(1) {
                    println!("{}", char);
                    if char.is_ascii_digit() {
                        number.push('.');
                        self.next();
                        number.push(char);
                        self.next();

                        end = loop {
                            if let Some((i, char)) = self.peek() {
                                match char {
                                    '0'..='9' => {
                                        self.next();
                                        number.push(char)
                                    }
                                    other if other.is_xid_start() => {
                                        return Err(LexerError::InvalidCharacter(Span::single(i)));
                                    }
                                    _ => {
                                        // This works because the character before it will be 1 byte long (0..9 || -)
                                        break i;
                                    }
                                }
                            } else {
                                break idx + number.len();
                            }
                        }
                    } else if char.is_xid_start() {
                        return Err(LexerError::InvalidCharacter(Span::single(i)));
                    }
                }
            } else if char.is_ascii_alphabetic() {
                return Err(LexerError::LetterInNumber(Span::start_end(idx, i)));
            }
        }

        if number.contains('.') {
            let number = number
                .parse::<f64>()
                .map_err(|_| LexerError::FloatParseError(Span::start_end(idx, end)))?;
            Ok(Token::new(TokenKind::Float(number), idx, end))
        } else {
            let number = number
                .parse::<i64>()
                .map_err(|_| LexerError::IntParseError(Span::start_end(idx, end)))?;
            Ok(Token::new(TokenKind::Int(number), idx, end))
        }
    }

    fn compute_string(&mut self, start: usize) -> LexerResult<Token> {
        let mut str = String::new();
        while let Some((idx, char)) = self.next() {
            match char {
                '\\' => {
                    let (_, next_char) = self.next().ok_or(LexerError::UnexpectedEOF)?;
                    match next_char {
                        '"' => str.push('\"'),
                        'n' => str.push('\n'),
                        'r' => str.push('\r'),
                        't' => str.push('\t'),
                        '0' => str.push('\0'),
                        '\\' => str.push('\\'),
                        _ => {
                            return Err(LexerError::InvalidEscapeCharacter(Span::start_end(
                                idx,
                                idx + 1,
                            )));
                        }
                    }
                }
                '"' => {
                    return Ok(Token::new(TokenKind::String(str), start, idx));
                }
                other => str.push(other),
            }
        }
        Err(LexerError::UnexpectedEOF)
    }
}

#[derive(Debug)]
pub enum LexerError {
    UnexpectedEOF,
    IntParseError(Span),
    FloatParseError(Span),
    InvalidCharacter(Span),
    InvalidEscapeCharacter(Span),
    LetterInNumber(Span),
}

impl CompilerError for LexerError {
    fn span(&self) -> Span {
        match self {
            LexerError::UnexpectedEOF => Span::dummy(),
            LexerError::IntParseError(span) => *span,
            LexerError::FloatParseError(span) => *span,
            LexerError::InvalidCharacter(span) => *span,
            LexerError::InvalidEscapeCharacter(span) => *span,
            LexerError::LetterInNumber(span) => *span,
        }
    }

    fn message(&self) -> String {
        match self {
            LexerError::UnexpectedEOF => "Unexpected EOF".to_string(),
            LexerError::IntParseError(_) => "Error parsing integer literal".to_string(),
            LexerError::FloatParseError(_) => "Error parsing float literal".to_string(),
            LexerError::InvalidCharacter(_) => "Invalid character".to_string(),
            LexerError::InvalidEscapeCharacter(_) => "Invalid escape character".to_string(),
            LexerError::LetterInNumber(_) => "Invalid letter in number".to_string(),
        }
    }

    fn note(&self) -> Option<String> {
        None
    }
}
