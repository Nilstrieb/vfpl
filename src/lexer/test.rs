use crate::lexer::tokens::Token;
use crate::lexer::{Lexer, LexerResult};
use crate::Session;

fn lex(code: &str) -> LexerResult<Vec<Token>> {
    Lexer::new(&code.to_lowercase(), Session::test_session()).compute_tokens()
}

#[test]
fn one_number() {
    insta::assert_debug_snapshot!(lex("1"));
}

#[test]
fn leading_zeros() {
    insta::assert_debug_snapshot!(lex("000012"));
}

#[test]
fn minus_number() {
    insta::assert_debug_snapshot!(lex("-12"));
}

#[test]
fn minus_with_leading_zeros() {
    insta::assert_debug_snapshot!(lex("-0000012"));
}

#[test]
fn alphabetical_char_in_number() {
    insta::assert_debug_snapshot!(lex("12a"));
}

#[test]
fn decimal() {
    insta::assert_debug_snapshot!(lex("13.35253"));
}

#[test]
fn decimal_with_leading_zeros() {
    insta::assert_debug_snapshot!(lex("000001.15"));
}

#[test]
fn decimal_with_minus_and_leading_zeros() {
    insta::assert_debug_snapshot!(lex("-00000012.23"));
}

#[test]
fn decimal_with_alphabetical_char() {
    insta::assert_debug_snapshot!(lex("00001.as244"));
}

#[test]
fn please_keyword() {
    insta::assert_debug_snapshot!(lex("please"));
}

#[test]
fn string() {
    insta::assert_debug_snapshot!(lex("\"Hello all\""));
}

#[test]
fn string_with_escape_chars() {
    insta::assert_debug_snapshot!(lex(r#""What is a \"unit test\"?""#));
}

#[test]
fn string_no_end() {
    insta::assert_debug_snapshot!(lex("\"Hahaha this is going to destroy it"));
}

#[test]
fn should_be_ident() {
    insta::assert_debug_snapshot!(lex("please hugo"));
}

#[test]
fn all_keywords() {
    let tokens = [
        "As",
        "Add",
        "Sub",
        "Mul",
        "Div",
        "Modulo",
        "Initialize",
        "Variable",
        "With",
        "The",
        "Value",
        "Of",
        "Set",
        "To",
        "From",
        "By",
        "Take",
        "End",
        "Check",
        "Whether",
        "Then",
        "Do",
        "Otherwise",
        "Break",
        "Out",
        "This",
        "Create",
        "Function",
        "Parameter",
        "Parameters",
        "That",
        "Returns",
        "Call",
        "No",
        "Argument",
        "Arguments",
        "And",
        "Go",
        "Sleep",
        "Absent",
        "Null",
        "NoValue",
        "Undefined",
        "True",
        "False",
        "Does",
        "Has",
        "Is",
        "Not",
        "Have",
        "Greater",
        "Less",
        "Than",
        "Equal",
        "Or",
        "Repeat",
        "Return",
        "While",
    ]
    .map(lex);
    insta::assert_debug_snapshot!(tokens);
}

#[test]
fn normal_line_of_code() {
    insta::assert_debug_snapshot!(lex(
        "please initialize variable x as Integer with the value of 15.\n"
    ));
}

#[test]
fn normal_line_of_code_decimal() {
    insta::assert_debug_snapshot!(lex(
        "please initialize variable x as Integer with the value of 15.3.\n"
    ));
}
