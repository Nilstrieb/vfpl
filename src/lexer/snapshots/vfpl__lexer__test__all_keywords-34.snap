---
source: src/lexer/test.rs
expression: "lex(\"No\")"

---
Ok(
    [
        Token {
            span: Span {
                start: 0,
                end: 2,
            },
            kind: No,
        },
    ],
)
